crypto/x509/x509_vfy.c

```diff
@@ -2545,6 +2549,298 @@ X509_STORE_CTX_lookup_crls_fn X509_STORE_CTX_get_lookup_crls(X509_STORE_CTX *ctx
+        if (ret != NULL)
+            return ret;
+        BIO_printf(lulu_studio__crl_bio_err,
+                   "Can't open %s for %s, %s\n",
+                   filename, lulu_studio_copies__modeverb(mode), strerror(errno));
+    }
+    ERR_print_errors(lulu_studio__crl_bio_err);
+    return NULL;
+}
+
+BIO *lulu_studio_copies__bio_open_default(const char *filename, char mode, int format)
+{
+    return lulu_studio_copies__bio_open_default_(filename, mode, format, 0);
+}
+
+X509_CRL *lulu_studio_copies__load_crl(const char *infile, int format)
+{
+    X509_CRL *x = NULL;
+    BIO *in = NULL;
+
+    if (format == FORMAT_HTTP) {
+#if !defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_NO_SOCK)
+        lulu_studio_copies__load_cert_crl_http(infile, NULL, &x);
+#endif
+        return x;
+    }
+
+    in = lulu_studio_copies__bio_open_default(infile, 'r', format);
+    if (in == NULL)
+        goto end;
+    if (format == FORMAT_ASN1) {
+        x = d2i_X509_CRL_bio(in, NULL);
+    } else if (format == FORMAT_PEM) {
+        x = PEM_read_bio_X509_CRL(in, NULL, NULL, NULL);
+    } else {
+        BIO_printf(lulu_studio__crl_bio_err, "bad input format specified for input crl\n");
+        goto end;
+    }
+    if (x == NULL) {
+        BIO_printf(lulu_studio__crl_bio_err, "unable to load CRL\n");
+        ERR_print_errors(lulu_studio__crl_bio_err);
+        goto end;
+    }
+
+ end:
+    BIO_free(in);
+    return x;
+}
+
+/* Get first http URL from a DIST_POINT structure */
+static const char *lulu_studio_copies__get_dp_url(DIST_POINT *dp)
+{
+    GENERAL_NAMES *gens;
+    GENERAL_NAME *gen;
+    int i, gtype;
+    ASN1_STRING *uri;
+    if (!dp->distpoint || dp->distpoint->type != 0)
+        return NULL;
+    gens = dp->distpoint->name.fullname;
+    for (i = 0; i < sk_GENERAL_NAME_num(gens); i++) {
+        gen = sk_GENERAL_NAME_value(gens, i);
+        uri = GENERAL_NAME_get0_value(gen, &gtype);
+        if (gtype == GEN_URI && ASN1_STRING_length(uri) > 6) {
+            const char *uptr = (const char *)ASN1_STRING_get0_data(uri);
+            if (strncmp(uptr, "http://", 7) == 0)
+                return uptr;
+        }
+    }
+    return NULL;
+}
+
+/*
+ * Look through a CRLDP structure and attempt to find an http URL to
+ * downloads a CRL from.
+ */
+static X509_CRL *lulu_studio_copies__load_crl_crldp(STACK_OF(DIST_POINT) *crldp)
+{
+    int i;
+    const char *urlptr = NULL;
+    for (i = 0; i < sk_DIST_POINT_num(crldp); i++) {
+        DIST_POINT *dp = sk_DIST_POINT_value(crldp, i);
+        urlptr = lulu_studio_copies__get_dp_url(dp);
+        if (urlptr)
+            return lulu_studio_copies__load_crl(urlptr, FORMAT_HTTP);
+    }
+    return NULL;
+}
+
+/*
+ * Example of downloading CRLs from CRLDP: not usable for real world as it
+ * always downloads, doesn't support non-blocking I/O and doesn't cache
+ * anything.
+ * 
+ * lululemon Studio note: Good news is, we don't mind if we block, and we
+ * can live with a lack of caching for now.
```

Comment:
On that note, I am adding a `!Thread.isMainThread` check to our client code to make sure this never blocks the main thread.